std/random.d

```diff
@@ -1934,6 +1934,177 @@ A singleton instance of the default random number generator
     assert(rnd.take(3).sum > 0);
 }
 
+// OS RND code is based on mir.random, https://github.com/libmir/mir-random/blob/master/source/mir/random/engine/package.d
+
+version (FreeBSD)
+    version = UseFileRandom;
+version (linux)
+    version = UseFileRandom;
+
+version (Windows)
+{
+    import core.sys.windows.winbase, core.sys.windows.wincrypt,
+        core.sys.windows.windows, core.sys.windows.winerror,
+        std.windows.syserror;
+
+    struct WindowsRandom
+    {
+    private:
+        ULONG_PTR _hProvider;
+
+        void initialize() @trusted
+        {
+            // https://msdn.microsoft.com/en-us/library/windows/desktop/aa379886(v=vs.85).aspx
+            // For performance reasons, we recommend that you set the pszContainer
+            // parameter to NULL and the dwFlags parameter to CRYPT_VERIFYCONTEXT
+            // in all situations where you do not require a persisted key.
+            // CRYPT_SILENT is intended for use with applications for which the UI cannot be displayed by the CSP.
+            if (!CryptAcquireContextW(&_hProvider, null, null, PROV_RSA_FULL, CRYPT_VERIFYCONTEXT | CRYPT_SILENT))
+            {
+                if (GetLastError() == NTE_BAD_KEYSET)
+                {
+                    // Attempt to create default container
+                    wenforce (CryptAcquireContextA(&_hProvider, null, null, PROV_RSA_FULL,
+                    CRYPT_NEWKEYSET | CRYPT_SILENT), "Could not acquire crypt context");
+                }
+                else
+                {
+                    wenforce (false, "Could not acquire crypt context");
+                }
+            }
+
+            popFront();
+        }
+
+    public:
+        enum bool isUniformRandom = true;
+        enum empty = false;
+
+        ulong front;
+
+        void popFront() @trusted nothrow @nogc
+        {
+            if (!CryptGenRandom(_hProvider, front.sizeof, cast(PBYTE) &front))
```

Comment:
Shouldn't we fetch more bytes?

https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptgenrandom

Reply:
You mean keepin it in a local buffer? I think we can't really return more data, as far as I know the random API expects front to be an integer type.

But what kind of size would be best here? The MSDN docs don't seem to provide that information?

BTW, that's the reason I feel uncomfortable doing all this crypto stuff, you have to double check every single detail and you might still miss something: For example, I'm not sure if we can/should just keep data which is used for cryptographic purposes in "unprotected" buffers (e.g. buffers which can be swapped out). But I guess it's the same for front already and the current RND generator API / range API doesn't allow more safe interfaces (where we only write to a user-specified destination).

Reply:
BTW, caching anything locally in the struct also means that everything returned by `osRNDGen` needs to be somehow by reference, to ensure we never return duplicated data. It could return by reference, but then it's not possible to store it in a local variable.... `RefCounted` seems a bit overkill. Does Unique work well enough for such cases nowadays? Last time I tried, Unique ranges were often not working it many parts of phobos...